### PR TITLE
BUG [table]: fix a bug where column names would be lost when instanciating Table from a list of Row objects

### LINKED ISSUE #5923: Column names not preserved when initializing Table from a list of Rows
The following code:

```python
import numpy as np
from astropy.table import Table

t1 = Table()
t1['a'] = [1,2,3]
t1['b'] = [2.,3.,4.]

print(t1)

rows = []
for row in t1:
    rows.append(row)

t2 = Table(rows=rows)

print(t2)
```

illustrates that column names are lost:

```
 a   b 
--- ---
  1 2.0
  2 3.0
  3 4.0
col0 col1
---- ----
   1  2.0
   2  3.0
   3  4.0
```

Of course, this isn't the recommended way of getting a Table subset, but I think that it's nevertheless a bug?

### PR DESCRIPTION
**Pull request by neutrinoceros**
### Description

Fixes #5923
This is a somewhat inelegant minimal patch; I'll be happy to iterate over it to improve quality if requested !

ping @hamogu @taldcroft

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: astropy/table/table.py
Comment (pllim) [inline astropy/table/table.py:623]: Addition of new accepted type as input also feel like a feature, so I didn't mark this as backport but Table maintainers can always change that if they want.

Thanks!

**Comment by neutrinoceros** [comment]
There's one failure I didn't catch locally. I'll get back to fix it in the morning, most likely.

File: astropy/table/table.py
Comment (neutrinoceros) [inline astropy/table/table.py:235]: note to reviewers: I'm assuming that passing rows of different sizes would raise an exception sooner or later during initialization, but I think this function should be as resilient as possible, so raising that exception is not its responsibility.

File: astropy/table/tests/test_table.py
Comment (neutrinoceros) [inline astropy/table/tests/test_table.py:3359]: I used a noqa comment here to keep this intention of that line very clear. Otherwise, the linter would push for
`rows = list(t1)`, which is equivalent, but a bit more opaque in my opinion.

Comment (neutrinoceros) [reply astropy/table/tests/test_table.py:3359]: I could also create a list of `Rows` manually. Here I just took the example from @astrofrog for easier comparison between the test and its linked issue.

File: astropy/table/table.py
Comment (taldcroft) [reply astropy/table/table.py:624]: Initializing from a sequence of Row was always intended to work, so this is just documenting that correctly (and fixing a bug).

Comment (pllim) [reply astropy/table/table.py:624]: So, @taldcroft , should we backport?

**Comment by hamogu** [comment]
We should backport.

**Comment by hamogu** [comment]
Since I don't review many PRs, I don't know what I need tp do about that though. Set the backport label?

**Comment by pllim** [comment]
Change the milestone and set the backport label, yes. I already did it here. Thanks for the feedback!

File: astropy/table/table.py
Comment (taldcroft) [reply astropy/table/table.py:624]: Yes on the backport in concept, though to be honest I don't know where it will be backported. I thought there is only 6.0.1 which is the next bugfix release forward from 6.0.0.

Also, of course, I haven't reviewed the code yet!

**Comment by taldcroft** [comment]
@neutrinoceros - I did a little digging and noticed a solution that has a much smaller footprint, in fact it reduces the line count. Basically this relies on `Row` being mostly dict-like, though sadly `list(row)` gives the values not the keys. (If we could go back 11 years...).

This also fixes some old inelegant code for getting the column order which was written way before Python had an ordered dict. It's very easy now to get the column order as simply the order in which they appear in successive rows. The previous alphabetical sort was not very nice.

One slight unfortunate thing about this solution is that it could change the column ordering in code so it would count as an API change for the unusual circumstance where the first row does not have all the columns. So I think this would need to wait for the next release, but the original issue is 6 years old so it can wait a little longer.
```
diff --git a/astropy/table/table.py b/astropy/table/table.py
index ac5a06d791..65631d9b0a 100644
--- a/astropy/table/table.py
+++ b/astropy/table/table.py
@@ -205,9 +205,9 @@ def _get_names_from_list_of_dict(rows):
     if rows is None:
         return None
 
-    names = set()
+    names = dict()
     for row in rows:
-        if not isinstance(row, Mapping):
+        if not isinstance(row, (Mapping, Row)):
             return None
         names.update(row)
     return list(names)
@@ -1169,17 +1169,7 @@ class Table:
         MISSING = object()
 
         # Gather column names that exist in the input `data`.
-        names_from_data = set()
-        for row in data:
-            names_from_data.update(row)
-
-        if set(data[0].keys()) == names_from_data:
-            names_from_data = list(data[0].keys())
-        else:
-            names_from_data = sorted(names_from_data)
-
-        # Note: if set(data[0].keys()) != names_from_data, this will give an
-        # exception later, so NO need to catch here.
+        names_from_data = _get_names_from_list_of_dict(data)
 
         # Convert list of dict into dict of list (cols), keep track of missing
         # indexes and put in MISSING placeholders in the `cols` lists.
```

**Comment by neutrinoceros** [comment]
@taldcroft thank you so much, I agree your solution is much more elegant and it still passes the test, so I went ahead and replaced mine with it (with you as a "co-author") !

**Comment by taldcroft** [comment]
@neutrinoceros - the sphinx doctests are failing which highlights the API change I mentioned. So there are a couple of actions:
- Update the sphinx narrative docs to pass.
- Add a `15735.api.rst` file which describes the API change for the case of `list[Mapping | Row]` input data where the first row does not have all the columns.
- Add a unit test to catch this case, this update should have been failing unit tests originally!

I've changed the milestone to 6.1 and added the `API` label.

**Comment by neutrinoceros** [comment]
Thanks @taldcroft , I think I've addressed all your points !

**Comment by saimn** [comment]
Hmm API changes should go in 7.0 (with a deprecation warning if possible), not 6.1...

File: docs/changes/table/15735.api.rst
Comment (saimn) [inline docs/changes/table/15735.api.rst:28]: Changelog is usually limited to one line or a few ones, no examples... maybe a good hint that this deserves a whatnew section in 7.0 ?

**Comment by pllim** [comment]
As per Simon's recommendation, I moved milestone to 7.0, added label to request What's New entry, and turned this into draft so we don't merge before v6.1.

**Comment by pllim** [comment]
Re: https://github.com/astropy/astropy/pull/15735#issuecomment-1863471443

@saimn , I have not been super careful about API change and v6.1. Do we have to revert anything in https://github.com/astropy/astropy/pulls?q=is%3Apr+is%3Amerged+label%3A%22API+change%22+milestone%3Av6.1.0

**Comment by hamogu** [comment]
I think pushing this to 7.0 is going too far. See https://github.com/astropy/astropy/pull/15735#discussion_r1427881233
This was always intended to work -in my view  that fact that it did not is a bug that's fixed here. In that sense it's similar to https://github.com/astropy/astropy/pull/15774 - it breaks stuff if you relied on undocumented, uncodified behavior that the was not intended. In the same vain, I don't think a what'-s new is justified for this bug fix.

@taldcroft ?

**Comment by neutrinoceros** [comment]
>  it breaks stuff if you relied on undocumented, uncodified behavior that the was not intended.

@hamogu  not exactly: the behaviour that's changed here *was* documented, even though it wasn't ideal (and it's a side effect from @taldcroft's patch that I adopted, not it's primary goal)
In that sense, I agree that it makes sense to push it to 6.1 in this kind of minor breakage was acceptable (it sounds like they're not) or 7.0 otherwise.

My own patch was backward compatible, albeit a lot less elegant than @taldcroft , so I think I can propose a way forward that should be satisfactory to all parties:
- reinstate my own backward-compatible patch that *just* fixes the targeted bug (including a regression test) and move back this PR to 6.0.1
- open a separate PR with @taldcroft's cleanup patch, aimed at 7.0

@saimn I am not sure how to go about adding a deprecation warning in this case: ideally users should have some sort of control to opt-in the new behaviour before it becomes the default, so they're able to *dodge* the warning once they are aware of it. In this case in do not see how to provide such an interface without adding a flag to `Table.__init__`'s signature, but the flag itself would then need to be deprecated too. All in all, is it really worth it ? I think there's a case to be had that a clean (documented !)  break in 7.0 would help a lot with maintainability.

**Comment by saimn** [comment]
@hamogu - From what I understand, and @neutrinoceros confirms it above, there is both a bugfix (supporting list of Row) and an API change (ordering of the columns, which affects also instantiating from a list of dict). If the second part can be avoided then the bugfix can even go to 6.0.1, and maybe the API change can be delayed to 7.0, with a warning if possible.

@neutrinoceros - Not sure for the warning, it also depends if it makes sense to have an option to keep the old behavior. One option is to issue the warning when columns are sorted, stating that the default behavior will change in 7.0, and people can filter the warning if they want. Or add an option to choose the sort behavior which would allow to force the old behavior if it make sense to do so ?

**Comment by neutrinoceros** [comment]
> Or add an option to choose the sort behavior which would allow to force the old behavior if it make sense to do so ?

What's great about @taldcroft's patch is that it decreases complexity. If we were to support multiple behaviours then complexity would go back up again, and I'm honestly not sure it's worth it.

**Comment by taldcroft** [comment]
This is exactly the sort of situation I was thinking about in discussions for APE-21 which resulted in the statement about being pragmatic. **Historically, API changes like this have been common in minor releases**. Although we justified them by pointing to the LTS, in practice the project decided that very few people were actually using the LTS.

Breakage could only occur for a pretty unusual situation:
1. Initializing a table from a list of dict where the first row does not contain all of the columns.
2. Downstream code that depends on the column ordering.

The pragmatic part is accepting that this situation can occur but it is unlikely. Not only does (1) need to be true, but it needs to always be true for all input data sets. Basically the current code is somewhat randomly changing the column ordering based on an arbitrary test.

 It is true that we could test for situation (1), issue a warning and sort the columns alphabetically. But then users in that situation need to modify their code anyway to suppress the warning. And if they want to permanently keep alphabetical ordering they have to add code to do that sorting so it still works after 7.0, and then they are left with orphaned code to suppress the warning. Users that want the more natural ordering now would be out of luck unless we add some new configuration option. 

This is just a LOT of complexity for a very little gain.

**Comment by hamogu** [comment]
Like #15774 this sounds like space-bar heating: https://xkcd.com/1172/. I don't think we have to backport, but I suggest a 6.1 milestone. We also don't have to do the extra work of breaking this up into two PRs.

**Comment by taldcroft** [comment]
In my day job (Chandra Flight Director dealing directly with NASA) I deal with process control authority, A LOT. In this case, there is a delegation of authority:
- CoCo has ultimate authority for the release contents.
- CoCo does not have time to deal with each PR, so they delegate to the release managers.
- Release managers do not have time or (in many cases) the domain knowledge to assess each PR, so they delegate to the package maintainers.
- Package maintainers normally have authority to make a decision about the impact of an API change and the target milestone.

Only in unusual circumstances would the release managers or the CoCo override the recommendation of the package maintainer(s). Of course a healthy discussion is always encouraged!

**Comment by saimn** [comment]
Not sure I understand where this discussion is going but, sure in the end it's up to the maintainers to decide. 
But on the other side the change of philosophy of the release cycle is recent and it's easy to forget (e.g. #14977) that we decided to avoid breaking changes in minor versions. It's sometimes difficult to avoid, or sometimes justified to break something in order to fix a bug, fine. 
Here the first version was backward compatible, and creating a Table from a list of dict is pretty common I guess, so it seems reasonable to ask if we can find another way to proceed.

**Comment by neutrinoceros** [comment]
It's not up to me to decide and ultimately I'm happy to re-iterate as long as it takes to get everyone happy but I think that going with the patch as is and aiming it as 6.1 is a reasonable way forward and indeed falls into the "pragmatic" kind of low impact breakage that APE 21 defines as acceptable for a minor release.

**Comment by taldcroft** [comment]
@saimn - agreed to try avoiding API changes, I'm definitely 100% behind not breaking code. That said, in this case I believe the complexity that is needed to maintain perfect back-compatibility outweighs the potential impact. Creating Tables from list of dict is common, but having both (1) and (2) from above is not.

File: docs/changes/table/15735.api.rst
Comment (taldcroft) [reply docs/changes/table/15735.api.rst:28]: This is an interesting question since we need to communicate this API change effectively. Would this go in the What's New for 6.1? Is that too high-profile or just right?

In general, is there any technical or aesthetic reason to not have longer changelog entries with examples? I'm just wondering about the driver for the precedent.

Comment (saimn) [reply docs/changes/table/15735.api.rst:28]: Currently whatsnew is probably the best place. 
Changelog entries used to be restricted to one line per PR, then with towncrier we relaxed a bit this constraint (nobody verifies it), but the current format is not really adequate to have long explanations mixed with one-line entries.  Also because it all goes in one big file.
I suggested in the APE22 discussions that we could improve this, maybe with some inspiration from pandas (e.g. https://pandas.pydata.org/docs/whatsnew/v2.1.0.html), to have a better presentation of changes and more room to highlight important ones. Maybe something to discuss.

Comment (neutrinoceros) [reply docs/changes/table/15735.api.rst:28]: I'm confused, my understanding was that "what's new" pages were constructed from changelog sample files. Where should I put this note then ?

Comment (hamogu) [reply docs/changes/table/15735.api.rst:28]: THe what's new is hand-written: https://github.com/astropy/astropy/tree/main/docs/whatsnew

File: docs/whatsnew/6.1.rst
Comment (taldcroft) [inline docs/whatsnew/6.1.rst:36]:
```suggestion
Ordering of table columns constructed from rows
===============================================

The column order in a `~astropy.table.Table` constructed from a `list` or rows (`dict`
or `~astropy.table.Row`) may change using ``astropy`` version ``6.1`` *if the 
first row has missing values*.

Before ``6.1``, the column ordering was determined from the first row if it contained
values for all the columns, or by sorting the final column names alphabetically if it 
did not. Starting with ``6.1``, columns are always added in the order they appear 
when iterating over  the `list` of rows.
```

File: docs/whatsnew/6.1.rst
Comment (taldcroft) [inline docs/whatsnew/6.1.rst:42]: A list of rows as `list[dict | Row]` can be passed as the first argument, and that is probably the most common. The `rows=..` is only needed for a `list[list]` input. But we can highlight that both work.
```suggestion
    >>> t = Table(data)  # or Table(rows=data), which is equivalent
```

File: docs/whatsnew/6.1.rst
Comment (taldcroft) [inline docs/whatsnew/6.1.rst:38]: As a general rule, do not make a single long sentence containing multiple code blocks. Instead make a short sentence or clause in front of each one.
```suggestion
For example, create a table as shown below:
```

File: docs/whatsnew/6.1.rst
Comment (taldcroft) [inline docs/whatsnew/6.1.rst:43]: Sphinx testing will fail because this does not print anything.
```suggestion
```

File: docs/whatsnew/6.1.rst
Comment (taldcroft) [inline docs/whatsnew/6.1.rst:46]:
```suggestion
Before ``6.1`` the table would look like this:
```

File: docs/whatsnew/6.1.rst
Comment (taldcroft) [inline docs/whatsnew/6.1.rst:54]:
```suggestion
Starting with ``6.1`` the table would instead look like this:
```

File: docs/whatsnew/6.1.rst
Comment (taldcroft) [inline docs/whatsnew/6.1.rst:40]:
```suggestion
    >>> from astropy.table import Table
    >>> data = [{'b': 10, 'c': 7, },
```

**Review by taldcroft — CHANGES_REQUESTED** [review]
@neutrinoceros - this is converging! I made a number of suggestions on the docs. The code and tests look fine. I think the CI failures are just doc tests (at least from the first one) but you should check them all.

File: docs/changes/table/15735.api.rst
Comment (pllim) [reply docs/changes/table/15735.api.rst]: If this needs What's New again, please re-apply the What's New needed label. Thanks!

**Comment by neutrinoceros** [comment]
rebased to resolve a merge conflict

**Comment by neutrinoceros** [comment]
rebased again to resolve conflicts. This time I also squashed my commits to minimize the number of conflicts I need to resolve in a single rebase.

**Comment by neutrinoceros** [comment]
rebased again to repel the bot.

**Comment by taldcroft** [comment]
@neutrinoceros - getting really close but RTD is failing, ugh. Apart from that I think this is approvable.

File: docs/whatsnew/7.0.rst
Comment (pllim) [inline docs/whatsnew/7.0.rst:256]:
```suggestion
Before ``7.0`` the table would look like this::
```

File: docs/whatsnew/7.0.rst
Comment (pllim) [inline docs/whatsnew/7.0.rst:264]:
```suggestion
Starting with ``7.0`` the table would instead look like this::
```

**Review by pllim — COMMENTED** [review]
Maybe the suggestions would fix the two `WARNING: Block quote ends without a blank line; unexpected unindent` warnings?

**Comment by neutrinoceros** [comment]
Thanks ! I included suggestions from @pllim and rebased again 🤞🏻

**Comment by pllim** [comment]
I restarted devdeps and mpldev, in case failures were transient. 🤞

**Comment by neutrinoceros** [comment]
@taldcroft last try before feature freeze ? 🤞🏻

**Comment by taldcroft** [comment]
Sorry, I'll have a look shortly.

**Review by taldcroft — APPROVED** [review]
Looks good!

**Pull request merged**
